tests/automated/1_program.py

```diff
@@ -13,13 +14,40 @@ def check_signature(p):
 
 @test_all_pandas
 @panda_connect_and_init(full_reset=False)
-def test_a_recover(p):
+def test_a_known_bootstub(p):
+  KNOWN_H7_BOOTSTUB_FN = os.path.join(BASEDIR, "tests", "automated", "known_bootstub", "bootstub.panda_h7.bin")
+  KNOWN_BOOTSTUB_FN = os.path.join(BASEDIR, "tests", "automated", "known_bootstub", "bootstub.panda.bin")
+  timeout = 30
+
+  p.reset(enter_bootstub=True)
+  p.reset(enter_bootloader=True)
+
+  t_start = time.time()
+  while len(PandaDFU.list()) == 0:
+    print("waiting for DFU...")
+    time.sleep(0.1)
+    if timeout is not None and (time.time() - t_start) > timeout:
+      return False
+
+  dfu = PandaDFU(PandaDFU.st_serial_to_dfu_serial(p._serial, p._mcu_type))
+  fn = KNOWN_H7_BOOTSTUB_FN if p._mcu_type == MCU_TYPE_H7 else KNOWN_BOOTSTUB_FN
```

Comment:
this is exactly the kind of thing we need to clean up. this is now done in the panda python lib, pandad, and here. the panda python lib might not even respect the filename you pass in: https://github.com/commaai/panda/blob/508ee90f8ecf1ce70fc9d002bbc14b47a50a459a/python/__init__.py#L340

Reply:
the panda python lib is a pretty high level library, but we don't consistently treat it that way

Reply:
agree about panda lib but don't understand how it is related to test. I am selecting known good bootstub file and not passing it to panda lib but flashing it.